### DISCUSSION TITLE
Creating new experiments with a new repo

### DISCUSSION THREAD

**Discussion by sroy729** [sarchlab/mgpusim#66] [General]
I want to use MGPUsim for my work.
while following the documentation it says it's better to create a git repo for running your experiments, so I tried to re-create for fir benchmark
Below are the steps I did for running my own benchmark

```
mkdir experiments
cd experiments
go mod init github.com/sroy729/mgpusim-experiments
cat main.go
package main
import (
              ...
)
func main() {
	fmt.Println("trying to run fir benchmark")
	runner := runner.Runner{}
	runner.Init()

	benchmark := fir.NewBenchmark(runner.GPUDriver)
	benchmark.Length = 4096
	runner.Benchmark = benchmark
	runner.Run()
}
```
for the import part, I have tried the following thing but all of them resulted in error
`"github.com/sarchlab/mgpusim/v3"`
or
```
github.com/sarchlab/mgpusim/v3/benchmarks/heteromark/fir
github.com/sarchlab/mgpusim/v3/samples/runner
```
As mentioned in the documentation, but after `go mod tidy` I got the error saying unable to find these. 

Can you tell me what are the minimum thing to do after creating a new repo to get a basic benchmark running.
Including what imports need to be done to create an independent repo for my experiments,

I think the problem is the runner and everything are not exported as a module. Can you help me with this

**Comment by Yzgll** [comment]
Do you want to add a new benchmark?

**Reply by sroy729** [reply]
Thank you for replying so fast. 
No, I don't want to add a new benchmark. I am using the existing benchmark only. I want to create my own custom repo containing all the experiments with different configurations, I was following your ISCA' 20 tutorial on YT
for example, my repo will look something like this 
```
experiments              // git repo 
       |-experiment 1  //with vanilla conguration of GPU the subdirectory will contain all the simulation results
               |- fir 
                     |- fir.go   //, but i think  the import statements need to be changed
       |-experiment 2   //change configuration of GPU and similarly subdirectory will contain all the simulation results
```
Something like the above.(do I need to clone MGPUsim here  ?)
In the documentation, you mentioned creating a new repository for each experiment and then copying files from `samples/experiment`, which is not there.
In the ISCA tutorial, you showed the experiments that you are running, pcie3 and pcie4, are in separate folders, but I cannot see the `main.go` file, which will have import statements.

I wanted to know what are the import statements to use in the new repo `main.go` file I created so that it can automatically get and compile the simulator and give the executable

**Comment by syifan** [comment]
I have created a minimal repo to demonstrate. There are a few APIs that are different from the ISCA tutorial time, but the structure is the same. Can you try to clone my repo and see if it can run in your environment? I tried it in my environment, and it works. There is no need to clone MGPUSim if you do not want to modify it. 

https://github.com/syifan/mgpusim_experiments

**Comment by sroy729** [comment]
Thanks it's working at my end also. I think I need to add your `go.mod` and `go.sum` file rather than generating one using go build initallly.
Also if let's say I need to increase the CU in some other experiment, do I just copy the `samples/runner/r9nanobuilder.go` in exp1 folder location and edit the configuration file. Will it run with update config or do i need to work with some other files too.

**Reply by syifan** [reply]
You will need to copy the whole runner directory. Go uses packages (code within a folder) as the minimum distribution/reuse unit. After copying the runner directory, do not forget to change the import lines in your main file.

**Reply by sroy729** [reply]
after doing that I am facing the following error
```
# github.com/syifan/mgpusim_experiments/exp2/runner
runner/timingplatform.go:186:3: gpuDriverBuilder.WithEngine(b.engine).WithPageTable(pageTable).WithLog2PageSize(b.log2PageSize).WithGlobalStorage(b.globalStorage).WithD2HCycles undefined (type "github.com/sarchlab/mgpusim/v3/driver".Builder has no field or method WithD2HCycles)
```
I commented out the lines and the build was successful. Why is this error popping up

**Reply by syifan** [reply]
This is because the runner code in the v3 branch is slightly newer than the most recent release v3.0.1. I have made a new release v3.0.2. If you change your go.mod file to update to the new version, it should work. Otherwise, commenting out the line should also not be a big problem.

**Reply by sroy729** [reply]
thank you

**Comment by sroy729** [comment]
How to get Daisen to work with it, I am running the following command 

./exp2 -timing

Which shows the following output
```
trying to run fir benchmark
Monitoring simulation with http://localhost:36745
```
But going to that URL, it does not work.

I tried to cloning the akita and go into the daisen directory and do a` go build` to generate a`daisen` executable 
After that I ran `./daisen -mysql <location of sql file generated during exp2> ` (I did get the ./exp to run with -trace-viz to get the db file ) but no success. The following is the output 
```
panic: trace username is not set, use environment variable AKITA_TRACE_USERNAME to set it.

goroutine 1 [running]:
github.com/sarchlab/akita/v3/tracing.(*dbConnection).getCredentials(0xc0000ae2a0)
	/home/roy_linux/mgpusim/sifan/mgpusim_experiments/akita/tracing/mysql.go:347 +0x15d
github.com/sarchlab/akita/v3/tracing.(*dbConnection).init(0xc0000ae2a0, {0x7fffbadd9f04?, 0xc000097f30?})
	/home/roy_linux/mgpusim/sifan/mgpusim_experiments/akita/tracing/mysql.go:340 +0x45
github.com/sarchlab/akita/v3/tracing.(*MySQLTraceReader).Init(...)
	/home/roy_linux/mgpusim/sifan/mgpusim_experiments/akita/tracing/mysql.go:173
main.connectToDB()
	/home/roy_linux/mgpusim/sifan/mgpusim_experiments/akita/daisen/server.go:74 +0x1ca
main.startServer()
	/home/roy_linux/mgpusim/sifan/mgpusim_experiments/akita/daisen/server.go:66 +0xf
main.main()
	/home/roy_linux/mgpusim/sifan/mgpusim_experiments/akita/daisen/server.go:36 +0x45
```

I tried `set -x AKITA_TRACE_USERNAME syifan` just following the MGPUSIM HPCS demo, but no luck.

Can you point me to the documentation to enable me to use the visualization tool so that when `Monitoring simulation with http://localhost:36745` comes, I can directly go to the URL?

**Reply by Yzgll** [reply]
You need to add "-trace-vis-db-file" at runtime. There are a number of parameters that can be added to the runtime, which you can take a look at.https://github.com/sarchlab/mgpusim/blob/v3/samples/runner/flag.go.     There is another paper on Daisen that you can also look at that will be very helpful in your understanding of Daisen's diagram. Although there are still some things I don't understand after reading it.HAHA

**Reply by sroy729** [reply]
Thanks I wil have a look at it. In the previous question I passed the flag incorrectly it should be `sqlite` rather than `mysql`. I am now able to see the Daisen plots..

**Reply by syifan** [reply]
BTW, the monitoring link allows you to monitor the simulation in real-time. It stopped working because the simulation terminated too early. If you try something like `./fir -timing -length=10485760`, it should work.
